From 0502c5847039fb62069f7e91ec47b7be0a08e888 Mon Sep 17 00:00:00 2001
From: Heinz Wrobel <Heinz.Wrobel@nxp.com>
Date: Tue, 16 Feb 2021 18:46:00 +0200
Subject: [PATCH 80/83] pcie: S32 Gen 1 and S32G device trees had bad PCIe
 mappings

There were overlaps and the memory map range was not available.
Now the config space is put at the very end with I/O space before.
This eliminates overlaps and permits best alignment for the memory
ranges.

Issue: ALB-6468

Signed-off-by: Heinz Wrobel <Heinz.Wrobel@nxp.com>
Signed-off-by: Ionut Vicovan <Ionut.Vicovan@nxp.com>
---
 arch/arm/dts/fsl-s32-gen1.dtsi | 20 ++++++++++++++------
 arch/arm/dts/fsl-s32g274a.dtsi | 20 ++++++++++++++------
 2 files changed, 28 insertions(+), 12 deletions(-)

diff --git a/arch/arm/dts/fsl-s32-gen1.dtsi b/arch/arm/dts/fsl-s32-gen1.dtsi
index 959e7f9393..4a72cfb340 100644
--- a/arch/arm/dts/fsl-s32-gen1.dtsi
+++ b/arch/arm/dts/fsl-s32-gen1.dtsi
@@ -438,8 +438,12 @@
 
 	pcie0: pcie@40400000 {
 		compatible = "fsl,s32gen1-pcie";
-		reg = <0x00 0x40400000 0x0 0x80000   /* dbi registers */
-		       0x58 0x00000000 0x0 0x20000>; /* configuration space */
+		reg =
+			<0x00 0x40400000 0x0 0x80000   /* dbi registers */
+			/* configuration space, 4KB each for cfg0 and cfg1
+			 * at the end of the outbound memory map
+			 */
+			0x5f 0xffffe000 0x0 0x00002000>;
 		reg-names = "dbi", "config";
 		#address-cells = <3>;
 		#size-cells = <2>;
@@ -451,10 +455,14 @@
 
 		bus-range = <0x0 0xff>;
 		ranges =
-			/* downstream I/O */
-			<0x81000000 0x0 0x00000000 0x58 0x00003000 0x0 0x00010000
-			/* non-prefetchable memory */
-			 0x82000000 0x0 0x00013000 0x58 0x00013000 0x0 0x40000000>;
+			/* downstream I/O, 64KB and aligned naturally just before
+			 * the config space to minimize fragmentation
+			 */
+			<0x81000000 0x0 0x00000000 0x5f 0xfffe0000 0x0 0x00010000
+			/* non-prefetchable memory, with best case size
+			 * and alignment
+			 */
+			 0x82000000 0x0 0x00000000 0x58 0x00000000 0x7 0xfffe0000>;
 		status = "disabled";
 	};
 
diff --git a/arch/arm/dts/fsl-s32g274a.dtsi b/arch/arm/dts/fsl-s32g274a.dtsi
index ddd0c614c4..9aa8a28c51 100644
--- a/arch/arm/dts/fsl-s32g274a.dtsi
+++ b/arch/arm/dts/fsl-s32g274a.dtsi
@@ -58,8 +58,12 @@
 
 	pcie1: pcie@44100000 {
 		compatible = "fsl,s32gen1-pcie";
-		reg = <0x00 0x44100000 0x0 0x80000   /* dbi registers */
-		       0x48 0x00000000 0x0 0x20000>; /* configuration space */
+		reg =
+			<0x00 0x44100000 0x0 0x80000   /* dbi registers */
+			/* configuration space, 4KB each for cfg0 and cfg1
+			 * at the end of the outbound memory map
+			 */
+			0x4f 0xffffe000 0x0 0x00002000>;
 		reg-names = "dbi", "config";
 		#address-cells = <3>;
 		#size-cells = <2>;
@@ -70,10 +74,14 @@
 		link-speed = <3>; /* supports Gen3 speed */
 		bus-range = <0x0 0xff>;
 		ranges =
-			/* downstream I/O */
-			<0x81000000 0x0 0x00000000 0x48 0x00020000 0x0 0x00010000
-			/* non-prefetchable memory */
-			0x82000000 0x0 0x40000000 0x48 0x40000000 0x0 0x40000000>;
+			/* downstream I/O, 64KB and aligned naturally just before
+			 * the config space to minimize fragmentation
+			 */
+			<0x81000000 0x0 0x00000000 0x4f 0xfffe0000 0x0 0x00010000
+			/* non-prefetchable memory, with best case size
+			 * and alignment
+			 */
+			 0x82000000 0x0 0x00000000 0x48 0x00000000 0x7 0xfffe0000>;
 		status = "disabled";
 	};
 
-- 
2.17.1

